### PR TITLE
Composable refactor

### PR DESCRIPTION
**Pull request by GavinHuttley**
Nearly finalised porting writers to new decorator approach

### CODE REVIEW

File: src/cogent3/app/io.py [hunk 1]
```diff
@@ -351,7 +358,7 @@ def main(self, path: IdentifierType) -> TabularType:
 
 
 @define_app(app_type=WRITER)
-class write_tabular:
+class write_tabular(_checkpointable):
```
Comment (Nick-Foto) [inline src/cogent3/app/io.py:361]: Here we mixed decorator approach and inheritance. Do we expect other users to use `_checkpointable` by inheritance or will we change it in the future?

File: tests/test_app/test_dist.py [hunk 2]
```diff
@@ -213,7 +216,6 @@ def test_composes_with_write_tabular(self):
             dist_calc = dist_app.fast_slow_dist(distance="hamming", moltype="protein")
             _ = dist_calc + writer
```
Comment (Nick-Foto) [inline tests/test_app/test_dist.py:217]: Do we always need have an assignment for result of composable apps? or 
Can we write this expression without assignment like this: `app1+app2` without assigning result to a variable? I imagine `app1+app2` like calling two functions: `function1(function2())`

File: src/cogent3/app/io.py (hunk 1 above)
Comment (GavinHuttley) [reply src/cogent3/app/io.py:361]: change in future

File: tests/test_app/test_dist.py (hunk 2 above)
Comment (GavinHuttley) [reply tests/test_app/test_dist.py:217]: I prefer to do the assignment but it's possible to do what you've said

**Pull request merged**